cartographer_ros/cartographer_ros/offline_node_main.cc

```diff
@@ -257,6 +257,17 @@ void Run(const std::vector<string>& bag_filenames) {
     }
 
     bag.close();
+
+    if (FLAGS_keep_running) {
```

Comment:
Ouch, now I see that this approach is flawed, especially with multi trajectory.

> What does this help? No more data is fed into Cartographer, so it will never do the final optimization, so your map will never converge to the final result.

That is true. However, it still _is_ optimizing in the background, right? And if it finishes everything, trajectory finalization is effectively done, and calling it is a no-op.

> I think you have to add hooks into the MapBuilder and maybe the sparse pose graph/constraint builder to give you an update on the state while it is optimizing.

But `FinalizeTrajectory()` blocks the main thread, which should be spinning.

I would be happiest if we could just do this, launch an `AsyncSpinner` and forget about it:
https://github.com/googlecartographer/cartographer_ros/compare/master...larics:keep-waiting-async

However, I am getting crashes. Would one of you be willing to join me on Hangouts, for example on Tuesday or Wednesday next week? I have captured the crash in a reversible debugger, and I am certain that you'll spot right away what's going on.

If that fails, I'll look into expanding this approach.